kpi/utils/query_parser/query_parser.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+
+from django.conf import settings
+from django.db.models import Q
+
+from .canopy_autogenerated_parser import parse as grammar_parse, ParseError
+
+"""
+This is a utility for parsing a Boolean, Whoosh-like query string and
+translating it into a Django Q object, which can then be used to filter a
+queryset in the ORM.
+
+Syntax examples:
+    * `name:term` returns any object whose `name` field exactly matches
+        `term` (case sensitive)
+    * `owner__username=meg` traverses the `owner` relationship, returning
+        any object whose `owner` field matches an object whose `username` field
+        exactly matches `meg` (case sensitive)
+    * `color:orange NOT (type:fruit OR type:chair)` returns anything
+        whose color is orange so long as it is not a fruit or a chair. AND, OR,
+        and NOT operators are supported. They must be in ALL CAPS.
+
+Special notes:
+    * If no field is specified in the query,
+        `settings.SEARCH_DEFAULT_FIELD_LOOKUP` is assumed. For example, if that
```

Comment:
Will it work whether `settings.SEARCH_DEFAULT_FIELD_LOOKUP` is `summary` and `summary` field is changed from JSONField to JSONBField?

Reply:
I guess if `summary` becomes a JSONB, we'd want `summary__icontains` to match on the stringified JSON representation, e.g. `summary__icontains:blah` would hopefully find an object whose summary looked like `{"deeply": [{"nested": "thing"}, {"blah": "blah"}]}`. I don't know whether or not it would work like that, but I'll try a test.

In the future, we can look at https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/search/ now that we're on a version of Django that supports it.

Reply:
@noliveleger, it seems like `icontains` does search through the entire JSON string by using `WHERE UPPER("jsonb_column"::text) LIKE UPPER('%blah%')`. Here's an example on `PerUserSetting`, just because it was a convenient model that had a `JSONBField`:
```
>>> uset = PerUserSetting()
>>> uset.user_queries = {"deeply": [{"nested": "thing"}, {"blah": "blah"}]}
>>> uset.save()
INSERT INTO "hub_perusersetting" ("user_queries",
                                  "name",
                                  "value_when_matched",
                                  "value_when_not_matched")
VALUES ('{"deeply": [{"nested": "thing"}, {"blah": "blah"}]}', 'INTERCOM_APP_ID', '', '') RETURNING "hub_perusersetting"."id" [4.86ms]
>>> PerUserSetting.objects.filter(user_queries__icontains='blah')
SELECT "hub_perusersetting"."id",
       "hub_perusersetting"."user_queries",
       "hub_perusersetting"."name",
       "hub_perusersetting"."value_when_matched",
       "hub_perusersetting"."value_when_not_matched"
FROM "hub_perusersetting"
WHERE UPPER("hub_perusersetting"."user_queries"::text) LIKE UPPER('%blah%')
LIMIT 21 [1.35ms]
<QuerySet [<PerUserSetting: INTERCOM_APP_ID>]>
```

Reply:
👍